target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh

```diff
@@ -29,8 +29,10 @@ platform_do_upgrade() {
 		;;
 	ampedwireless,ally-00x19k|\
 	ampedwireless,ally-r1900k)
-		if [ "$(fw_printenv --lock / -n bootImage 2>/dev/null)" != "0" ]; then
-			fw_setenv --lock / bootImage 0 || exit 1
+		if [ "$(grep -c bootImage=0 /dev/mtd1)" != "1" ]; then
```

Comment:
Please add a note to the commit message why you want to get rid of the uboot-envtools dependency.

Beside if that, it doesn't seem to me a good idea to do so. I mean, the uboot-envtools know about the inner structure of the u-boot env. With the change you are switching to hackish way of changing parameters stored in the uboot-env instead.

Furthermore, I'm in doubt that it works at all. To my knowledge, the u-boot env is protected by a checksum. After changing the bootImage parameter and without updating the checksum, u-boot should detect the env as corrupt and should use its internal env instead. U-boot should print a big warning about the corrupted env during start.

Reply:
I would love to continue requiring uboot-envtools, but as of 21.02.x, it is not included in the release initramfs images despite being listed in DEVICE_PACKAGES.  The ALLY devices require a 2-stage initial installation (initramfs first, then sysupgrade) and require a way to check and/or set the boot partition for sysupgrade.  fw_printenv/fw_setenv would be great but the dd/sed/mtd workaround works fine in my testing.

Reply:
You were correct, it detects the env as corrupted.  However it still works, because it reverts to the default environment, which uses the first boot image anyway.  So, just lucky there.
How do you recommend proceeding?  Is there a way I can mandate that uboot-envtools be embedded in these devices' initramfs?  Thanks.

Reply:
For initramfs builds the same rootfs will be used for all devices. You can add `uboot-envtools` to the `DEFAULT_PACKAGES` in target/linux/ramips/mt7621/target.mk and remove it from all `DEVICE_PACKAGES` for all mt7621 devices which do not use it with `-DEVICE_PACKAGES` Then it should be integrated in the default initramfs images.

Reply:
Thanks for the idea, I'll test this out.  It sounds fine to me as long as it doesn't end up increasing the initramfs image size too much for other mt7621 devices.

Reply:
This works in my testing, thanks.

Reply:
don't know what's going on with 21.02, but at least on the master branch uboot-envtools can be integrated into the initramfs. I just used it a few months ago. see 13308161788c98ae6cd48c22b13339fdb8c77130

Reply:
Strange, comparing the official initramfs and sysupgrade binary sizes, it seems that even the WiFi driver is not included.😳